{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ae82b393",
   "metadata": {},
   "source": [
    "# Models of Evolution\n",
    "\n",
    "1. Mutations\n",
    "2. Recombination\n",
    "3. Natural Selection\n",
    "4. Migration\n",
    "5. Mating\n",
    "6. Popluation Size\n",
    "\n",
    "This notebook explores fundamental evolutionary models and code that shows their impact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "206c637f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e4d92963",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fd24000",
   "metadata": {},
   "source": [
    "# Fundamentals of Genomics\n",
    "\n",
    "Evolution is driven through mutations accumulating in a chromosome.  A **chromosome** is a structure of nucleic acids and protein found in the nucleus of most living cells, carrying genetic information in the form of genes. For this lesson, we'll only evaluate the nucleic acids in a chromosome. In this exploration we will use *A* (adenine), *C* (cytosine), *G* (guanine), and *T* (thymine) nucleotides, also known as **bases**. Each position on a chromosome is called a **locus** and at each position the specific nucleotide is referred to as a **varient**.  A set of variants from a single chromosome ordered by locus is referred to as a **sequence**.  When multiple of the sequences are structured by order they become an **alignment** that allows you to see **alleles**, loci where more than one unique variant is present. \n",
    "\n",
    "\n",
    "## Alignment\n",
    "\n",
    "We will start by generating 8 identical sequence of 10 varients.  We will add a skew to the generation to bias the A and T nucleotides so that we get less of a random sequence.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c2f0e1e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0  1  2  3  4  5  6  7  8  9\n",
       "0  T  A  A  G  T  C  T  T  C  A\n",
       "1  T  A  A  G  T  C  T  T  C  A\n",
       "2  T  A  A  G  T  C  T  T  C  A\n",
       "3  T  A  A  G  T  C  T  T  C  A\n",
       "4  T  A  A  G  T  C  T  T  C  A\n",
       "5  T  A  A  G  T  C  T  T  C  A\n",
       "6  T  A  A  G  T  C  T  T  C  A\n",
       "7  T  A  A  G  T  C  T  T  C  A"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sets the seed so the random generation is the same each time\n",
    "np.random.seed(123)\n",
    "bases = ['A','C','G','T']\n",
    "\n",
    "# weights for our bases\n",
    "base_p = [0.4, 0.1, 0.1, 0.4] \n",
    "\n",
    "# create a single strand\n",
    "strand = np.random.choice(bases,10, p=base_p)\n",
    "\n",
    "# create alginment of 5 identical strands\n",
    "alignment = np.tile(strand,(8,1))\n",
    "\n",
    "#converting to pandas\n",
    "align_df = pd.DataFrame(alignment)\n",
    "align_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af29dda2",
   "metadata": {},
   "source": [
    "now we have an alignment where we have 8 strands (*n=8*), each strand has a length of 10 (*L=10*) and we have no sites with more than a single unique variant (*S=0*).   Let's now introduce a few **SNP** (single-nucleotide polymorphism), or a substitution of a single nucleotide at a specific position in the genome.   We'll add 5 SNPs at random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8d84cf1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0  1  2  3  4  5  6  7  8  9\n",
       "0  T  C  A  G  T  C  T  T  C  A\n",
       "1  T  A  A  G  T  C  T  T  C  A\n",
       "2  T  A  A  G  G  C  T  T  C  A\n",
       "3  T  A  A  G  T  A  T  T  C  A\n",
       "4  T  A  A  G  T  C  T  T  C  A\n",
       "5  T  A  A  G  T  C  T  T  G  A\n",
       "6  T  A  A  G  T  C  T  T  C  A\n",
       "7  T  A  A  G  T  C  C  T  C  A"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.random.seed(1)\n",
    "mut_df = align_df.copy()\n",
    "\n",
    "for i in range(5):\n",
    "    # creates a copy of the base list, generates a random strand and locus to mutate, \n",
    "    # removes the existing variant from the options, then picks a random choice of the bases left\n",
    "    mut = bases.copy()\n",
    "    strand = np.random.randint(0,8)\n",
    "    loci = np.random.randint(0,10)\n",
    "    existing_variant = mut_df[loci][strand]\n",
    "    mut.remove(existing_variant)\n",
    "    mut_df[loci][strand] = np.random.choice(mut)\n",
    "mut_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "704dd544",
   "metadata": {},
   "source": [
    "We've now have 5 loci with multiple Alleles (1,4,5,6,8) so now *S=5*.  Another form of mutation can be inserts and deletions. **Insertions** are when a strand acquires new bases that would generate a new loci not present in any other strand.  A **Deletions** is the opposite, or when a strand's base is removed at a loci. Both of these processes exert evolutionary impact and are often lumped togenter as \"InDels\" as there can be at times difficutly in discerning from which a new loci occurs. In both cases, the missingness of an allele (whether on the strand or on other strands) is represented with a period \".\".  We will now add in a deletion to our alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "906e419f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0  1  2  3  4  5  6  7  8  9\n",
       "0  T  C  A  G  T  C  T  T  C  A\n",
       "1  T  A  A  G  T  C  T  T  C  A\n",
       "2  T  A  A  G  G  C  T  T  C  A\n",
       "3  T  A  A  G  T  A  T  T  C  A\n",
       "4  T  A  A  G  T  C  T  T  C  A\n",
       "5  T  A  A  G  T  C  T  T  G  A\n",
       "6  T  A  .  G  T  C  T  T  C  A\n",
       "7  T  A  A  G  T  C  C  T  C  A"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.random.seed(123)\n",
    "strand = np.random.randint(0,8)\n",
    "loci = np.random.randint(0,10)\n",
    "mut_df[loci][strand] = '.'\n",
    "mut_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3aac458",
   "metadata": {},
   "source": [
    "At this point we now have 6 loci with multiple alleles *S=6*.  We can now calculate the allele frequency at each loci.  The allele frequency is the relative frequency of an allele (variant of a gene) at a particular locus in a population, expressed as a fraction or percentage.\n",
    "\n",
    "The frequency ($p$) of an allele *A* is the fraction of the number of copies ($n_A$) of the A allele and the population or sample size ($N$), so\n",
    "$$\n",
    "p = \\frac{n_A}{N}\n",
    "$$\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f2bf28a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_freq</th>\n",
       "      <td>{'T': 1.0}</td>\n",
       "      <td>{'A': 0.875, 'C': 0.125}</td>\n",
       "      <td>{'A': 0.875, '.': 0.125}</td>\n",
       "      <td>{'G': 1.0}</td>\n",
       "      <td>{'T': 0.875, 'G': 0.125}</td>\n",
       "      <td>{'C': 0.875, 'A': 0.125}</td>\n",
       "      <td>{'T': 0.875, 'C': 0.125}</td>\n",
       "      <td>{'T': 1.0}</td>\n",
       "      <td>{'C': 0.875, 'G': 0.125}</td>\n",
       "      <td>{'A': 1.0}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      0                         1                         2  \\\n",
       "0                     T                         C                         A   \n",
       "1                     T                         A                         A   \n",
       "2                     T                         A                         A   \n",
       "3                     T                         A                         A   \n",
       "4                     T                         A                         A   \n",
       "5                     T                         A                         A   \n",
       "6                     T                         A                         .   \n",
       "7                     T                         A                         A   \n",
       "allele_freq  {'T': 1.0}  {'A': 0.875, 'C': 0.125}  {'A': 0.875, '.': 0.125}   \n",
       "\n",
       "                      3                         4                         5  \\\n",
       "0                     G                         T                         C   \n",
       "1                     G                         T                         C   \n",
       "2                     G                         G                         C   \n",
       "3                     G                         T                         A   \n",
       "4                     G                         T                         C   \n",
       "5                     G                         T                         C   \n",
       "6                     G                         T                         C   \n",
       "7                     G                         T                         C   \n",
       "allele_freq  {'G': 1.0}  {'T': 0.875, 'G': 0.125}  {'C': 0.875, 'A': 0.125}   \n",
       "\n",
       "                                    6           7                         8  \\\n",
       "0                                   T           T                         C   \n",
       "1                                   T           T                         C   \n",
       "2                                   T           T                         C   \n",
       "3                                   T           T                         C   \n",
       "4                                   T           T                         C   \n",
       "5                                   T           T                         G   \n",
       "6                                   T           T                         C   \n",
       "7                                   C           T                         C   \n",
       "allele_freq  {'T': 0.875, 'C': 0.125}  {'T': 1.0}  {'C': 0.875, 'G': 0.125}   \n",
       "\n",
       "                      9  \n",
       "0                     A  \n",
       "1                     A  \n",
       "2                     A  \n",
       "3                     A  \n",
       "4                     A  \n",
       "5                     A  \n",
       "6                     A  \n",
       "7                     A  \n",
       "allele_freq  {'A': 1.0}  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for i in range(10):\n",
    "    #get values as dictionary\n",
    "    vals = mut_df[i].value_counts().to_dict()\n",
    "    \n",
    "    # get N\n",
    "    total_ct = sum(vals.values()) \n",
    "    \n",
    "    #iterate through each value and create a frequency dictionary\n",
    "    freq_dict = {j:vals[j]/total_ct for j in vals}\n",
    "    mut_df.loc['allele_freq',i] = str(freq_dict)\n",
    "mut_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87afa021",
   "metadata": {},
   "source": [
    "In the new row we can now clearly see the frequencies of each allele and the frequency of the missingness.  As we can see, even when there are multiple alleles, one tends to dominate within a popluation.  This is most common as allele variations, or mutations, are rarely advantageous and so they tend to stay in low frequencies.  We'll explore next some models of mutation to better understand why these alleles persist, even in low frequencies, and how they are introduced and spread."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ca4d695",
   "metadata": {},
   "source": [
    "## Wright-Fisher Model \n",
    "This is a common model for genetic drift.  **Genetic drift** is the stocastic change of alleles frequency in a population.  Drift differs from Natural Selection since Natural Selection is a deterministic change. Wright-fisher assumes the following\n",
    "1. The population remains at a constant size across generations.  This means that offsprings replace the parent. \n",
    "2. Individuals are diploid hermaphrodites.  This means that, if we include the sex chromosomes we have:\n",
    "    1. $2N$ Autosomes - non sex determining chromosomes\n",
    "    2. $0.5N$ mtDNA - the mitochondria come from only the mother \n",
    "    3. $1.5N$ X or Z - the chromosome in homogametic sex meaning half the population have 2 copies and the other half have 1 copy (e.g. Females in humans)\n",
    "    4. $0.5N$ Y or Q - the chromosome in heterogametic sex meaning only half the population has 1 copy (e.g. Males in humans)\n",
    "3. Mating is assumed to be random.\n",
    "\n",
    "This model is only really useful in species with non-overlapping periodic generations such as insects or plants. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b6888d9",
   "metadata": {},
   "source": [
    "### Generational Allele Frequency\n",
    "In the Wright-Fisher model, since each individual is a diploid, there is 2N chromosomes available.  Let's assume that at a loci, only 2 alleles can be present: $A_1$ and $A_2$. The count of $A_1$ can be represented as $i$ meaning we can represent the allele frequency at time $t$ as\n",
    "$$\n",
    "p_t = \\frac{i}{2N}\n",
    "$$\n",
    "With this, since there are only 2 alleles present at a location in the Wright-Fisher model, we can then assume that the allele frequency for $A_2$ at time $t$ is all of the remaning chromosomes or\n",
    "$$\n",
    "q_t = \\frac{2N-i}{2N}\n",
    "$$\n",
    "With this we can then derive a formula for quickly going from $p_t$ to $q_t$ using the following:\n",
    "\n",
    "$$p_t = \\frac{i}{2N}$$\n",
    "\n",
    "$$2Np_t = i$$\n",
    "\n",
    "$$q_t = \\frac{2N-i}{2N} = \\frac{2N-2Np_t}{2N} = \\frac{2N(1-p_t)}{2N} = 1-p_t$$\n",
    "\n",
    "$$q_t + p_t = 1$$\n",
    "\n",
    "This agrees with our logical interpretation since, if the total set of alleles (1) is the sum of all the allele frequencies, in a **biallelic** loci, or a loci with two alleles, this would be $q_t + p_t$\n",
    "\n",
    "\n",
    "In the Wright-Fisher model, the sampling of chromosomes for the next generation is the same as sampling from a binomial distribution leading to a mean $E$ and variance $Var$ of \n",
    "$$E(p_{t+1}) = p_t$$\n",
    "$$Var(p_{t+1}) = \\frac{p_t q_t}{2N}$$\n",
    "\n",
    "With this mean and variance, in a population that meets the Wright-Fisher model prediction, the alleles will drift until one of the alleles becomes **fixed** or reaches 100% frequency within a population. Let's observe taking locus 1 from our alignment as a example and running 100 generations 5 times.\n",
    "\n",
    "Let's start with Locus 1, and separate into 4 diploid individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ba9f3c18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_2</th>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_1</th>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_1</th>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_2</th>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_1</th>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_2</th>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele_freq</th>\n",
       "      <td>{'A': 0.875, 'C': 0.125}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    1\n",
       "1_1                                 C\n",
       "1_2                                 A\n",
       "2_1                                 A\n",
       "2_2                                 A\n",
       "3_1                                 A\n",
       "3_2                                 A\n",
       "4_1                                 A\n",
       "4_2                                 A\n",
       "allele_freq  {'A': 0.875, 'C': 0.125}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a copy of loci 1\n",
    "wf_drift = mut_df[[1]].copy()\n",
    "\n",
    "#generate a list that creates 4 diploid individuals and appends the allele frequency dict\n",
    "new_index = sorted([str(i+1) + '_' + str(j+1) for j in range(2) for i in range(4)] + [wf_drift.index[-1]])\n",
    "\n",
    "# rename the index\n",
    "wf_drift.index = new_index\n",
    "\n",
    "wf_drift"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7db5b67",
   "metadata": {},
   "source": [
    "conveniently we have the allele frequencies already calculated from before.  This allows us to quickly the expected mean and variance for our alleles. $p$ will refer to the base *A* and $q$ will refer to the base *C*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "bb5a867b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "p = 0.875\n",
      "q = 0.125\n",
      "var = 0.013671875\n"
     ]
    }
   ],
   "source": [
    "freq_dict = eval(wf_drift[1]['allele_freq'])\n",
    "p_t = freq_dict['A']\n",
    "q_t = freq_dict['C']\n",
    "var = (p_t*q_t)/(2*4)\n",
    "print(f'p = {str(p_t)}')\n",
    "print(f'q = {str(q_t)}')\n",
    "print(f'var = {str(var)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4557341b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
